File: sdk/typescript/src/services/files.ts
/**
 * Files service — upload, download, list, delete, and protect files.
 */
import { FastCMS } from '../client';

export interface FileRecord {
  id: string;
  filename: string;
  original_filename: string;
  mime_type: string;
  size: number;
  storage_path: string;
  storage_type: string;
  collection_name?: string;
  record_id?: string;
  field_name?: string;
  user_id?: string;
  is_thumbnail: boolean;
  protected: boolean;
  parent_file_id?: string;
  created: string;
  updated: string;
  url?: string;
}

export interface FileListResponse {
  items: FileRecord[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface FileUploadOptions {
  collectionName?: string;
  recordId?: string;
  fieldName?: string;
}

export class FilesService {
  constructor(private client: FastCMS) {}

  /**
   * Upload a file.
   */
  async upload(file: File | Blob, options: FileUploadOptions = {}): Promise<FileRecord> {
    const form = new FormData();
    form.append('file', file);
    if (options.collectionName) form.append('collection_name', options.collectionName);
    if (options.recordId) form.append('record_id', options.recordId);
    if (options.fieldName) form.append('field_name', options.fieldName);

    const response = await this.client.getAxios().post('/api/v1/files', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  }

  /**
   * List files with optional filters.
   */
  async list(options: {
    page?: number;
    perPage?: number;
    collectionName?: string;
    recordId?: string;
  } = {}): Promise<FileListResponse> {
    const params: Record<string, any> = {};
    if (options.page) params.page = options.page;
    if (options.perPage) params.per_page = options.perPage;
    if (options.collectionName) params.collection_name = options.collectionName;
    if (options.recordId) params.record_id = options.recordId;

    const response = await this.client.getAxios().get('/api/v1/files', { params });
    return response.data;
  }

  /**
   * Get file metadata by ID.
   */
  async getOne(fileId: string): Promise<FileRecord> {
    const response = await this.client.getAxios().get(`/api/v1/files/${fileId}`);
    return response.data;
  }

  /**
   * Build a download URL for a file.
   * Pass a thumb spec (e.g. "300x200", "300x200f") for on-demand thumbnails.
   */
  getDownloadUrl(fileId: string, thumb?: string, token?: string): string {
    const base = this.client.getConfig().baseUrl;
    let url = `${base}/api/v1/files/${fileId}/download`;
    const params: string[] = [];
    if (thumb) params.push(`thumb=${encodeURIComponent(thumb)}`);
    if (token) params.push(`token=${encodeURIComponent(token)}`);
    if (params.length) url += '?' + params.join('&');
    return url;
  }

  /**
   * Generate a short-lived access token for a protected file.
   */
  async getToken(fileId: string): Promise<{ token: string; expires_in: number }> {
    const response = await this.client.getAxios().post(`/api/v1/files/${fileId}/token`);
    return response.data;
  }

  /**
   * Set or unset the protected flag on a file.
   */
  async setProtected(fileId: string, protected_: boolean): Promise<FileRecord> {
    const response = await this.client.getAxios().patch(
      `/api/v1/files/${fileId}/protected`,
      null,
      { params: { protected: protected_ } }
    );
    return response.data;
  }

  /**
   * Delete a file.
   */
  async delete(fileId: string): Promise<void> {
    await this.client.getAxios().delete(`/api/v1/files/${fileId}`);
  }
}

File: sdk/python/fastcms/files.py
"""
Files service — upload, list, download URL, protect, delete.
"""
from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, BinaryIO

if TYPE_CHECKING:
    from .client import FastCMS, AsyncFastCMS


class FilesService:
    """Sync files service."""

    def __init__(self, client: "FastCMS"):
        self._client = client

    def upload(
        self,
        file: bytes | BinaryIO | Path,
        filename: str | None = None,
        *,
        collection_name: str | None = None,
        record_id: str | None = None,
        field_name: str | None = None,
    ) -> dict:
        """Upload a file. Returns the FileRecord dict."""
        if isinstance(file, Path):
            filename = filename or file.name
            file_bytes = file.read_bytes()
        elif isinstance(file, bytes):
            file_bytes = file
        else:
            file_bytes = file.read()

        fields: dict = {"file": (filename or "upload", file_bytes)}
        if collection_name: fields["collection_name"] = (None, collection_name)
        if record_id:       fields["record_id"] = (None, record_id)
        if field_name:      fields["field_name"] = (None, field_name)

        return self._client._request("POST", "/api/v1/files", files=fields)

    def list(
        self,
        *,
        page: int | None = None,
        per_page: int | None = None,
        collection_name: str | None = None,
        record_id: str | None = None,
    ) -> dict:
        params = {}
        if page:            params["page"] = page
        if per_page:        params["per_page"] = per_page
        if collection_name: params["collection_name"] = collection_name
        if record_id:       params["record_id"] = record_id
        return self._client._get("/api/v1/files", params=params)

    def get_one(self, file_id: str) -> dict:
        return self._client._get(f"/api/v1/files/{file_id}")

    def download_url(self, file_id: str, *, thumb: str | None = None, token: str | None = None) -> str:
        """Build the download URL for a file (no HTTP request made)."""
        base = self._client.base_url.rstrip("/")
        url = f"{base}/api/v1/files/{file_id}/download"
        params = []
        if thumb: params.append(f"thumb={thumb}")
        if token: params.append(f"token={token}")
        if params: url += "?" + "&".join(params)
        return url

    def get_token(self, file_id: str) -> dict:
        """Generate a short-lived download token for a protected file."""
        return self._client._post(f"/api/v1/files/{file_id}/token")

    def set_protected(self, file_id: str, protected: bool) -> dict:
        return self._client._request(
            "PATCH",
            f"/api/v1/files/{file_id}/protected",
            params={"protected": str(protected).lower()},
        )

    def delete(self, file_id: str) -> None:
        self._client._delete(f"/api/v1/files/{file_id}")


class AsyncFilesService:
    """Async files service (all methods are coroutines)."""

    def __init__(self, client: "AsyncFastCMS"):
        self._client = client

    async def upload(
        self,
        file: bytes | BinaryIO | Path,
        filename: str | None = None,
        *,
        collection_name: str | None = None,
        record_id: str | None = None,
        field_name: str | None = None,
    ) -> dict:
        if isinstance(file, Path):
            filename = filename or file.name
            file_bytes = file.read_bytes()
        elif isinstance(file, bytes):
            file_bytes = file
        else:
            file_bytes = file.read()

        fields: dict = {"file": (filename or "upload", file_bytes)}
        if collection_name: fields["collection_name"] = (None, collection_name)
        if record_id:       fields["record_id"] = (None, record_id)
        if field_name:      fields["field_name"] = (None, field_name)

        return await self._client._request("POST", "/api/v1/files", files=fields)

    async def list(
        self,
        *,
        page: int | None = None,
        per_page: int | None = None,
        collection_name: str | None = None,
        record_id: str | None = None,
    ) -> dict:
        params = {}
        if page:            params["page"] = page
        if per_page:        params["per_page"] = per_page
        if collection_name: params["collection_name"] = collection_name
        if record_id:       params["record_id"] = record_id
        return await self._client._get("/api/v1/files", params=params)

    async def get_one(self, file_id: str) -> dict:
        return await self._client._get(f"/api/v1/files/{file_id}")

    def download_url(self, file_id: str, *, thumb: str | None = None, token: str | None = None) -> str:
        base = self._client.base_url.rstrip("/")
        url = f"{base}/api/v1/files/{file_id}/download"
        params = []
        if thumb: params.append(f"thumb={thumb}")
        if token: params.append(f"token={token}")
        if params: url += "?" + "&".join(params)
        return url

    async def get_token(self, file_id: str) -> dict:
        return await self._client._post(f"/api/v1/files/{file_id}/token")

    async def set_protected(self, file_id: str, protected: bool) -> dict:
        return await self._client._request(
            "PATCH",
            f"/api/v1/files/{file_id}/protected",
            params={"protected": str(protected).lower()},
        )

    async def delete(self, file_id: str) -> None:
        await self._client._delete(f"/api/v1/files/{file_id}")
